### DISCUSSION TITLE
Progress Bar Only Updating once Method Complete

### DISCUSSION THREAD

**Discussion by renrutsirhc** [AvaloniaUI/Avalonia#5986] [Q&A] [answered]
Hi there,

I'm struggling to get something working and hoping someone can point me in the right direction. I have a ReactiveCOmmand which is called from a button that does some processing on some excel sheets. 

**Constructor for VM**
```
 Progress<ProgressReportModel> progress = new Progress<ProgressReportModel>();
            progress.ProgressChanged += OnProgressChanged;
            ProcessPerDiemsCommand = ReactiveCommand.CreateFromTask(() => ProcessPerDiemsAsync(progress));
```
VM
```
      private void OnProgressChanged(object? sender, ProgressReportModel e)
        {
            ProgressIncrement = (e.Progress * 100) / e.ProgressMax;
            ProgressText = e.CurrentStage;
        }
        private double _progressIncrement;

        public double ProgressIncrement
        {
            get { return _progressIncrement; }
            set => this.RaiseAndSetIfChanged(ref _progressIncrement, value);
        }

        private string _progressText;

        public string ProgressText
        {
            get { return _progressText; }
            set => this.RaiseAndSetIfChanged(ref _progressText, value);
        }
        public ReactiveCommand<Unit, Unit> ProcessPerDiemsCommand { get; }

        private async Task ProcessPerDiemsAsync(Progress<ProgressReportModel> progress)
        {
//Do some stuff, occasionally calling progress.Report(report); where report is a ProgressReportModel


        }
```

The code is all working fine - the problem is that the OnProgressChanged Event Handler is only called once the ReactiveCommand finishes running and so my progressbar which is bound to ProgressIncrement doesn't update until after the command is already finished - slightly defeating the point! Can anyone shed any light?

**Comment by maxkatz6** [comment, marked as answer]
I don't see anything wrong with this code. Problem might be inside of ProcessPerDiemsAsync method.
It also important to remember, that Progress invokes ProgressChanged on sync context on which Progress object was created (see [dotnet/runtime/Progress.cs#L34](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Progress.cs#L34)). So you need to ensure that you creates that object on UI thread.
If it didn't help, can you share minimal project to run and test?

**Reply by renrutsirhc** [reply]
Thanks Max. It turned out that the code in the ProcessPerDiemsAsync method was running synchronously and that this caused the problem. When made to run in a new thread using await Task.Run, this worked fine. Thanks for your help!
